python,misc,stdlib: addressing mypy lints

AnthonyMichaelTDM opened the pull request:
This pr includes various type fixes, small refactors, import updates, etc. that aim to reduce the number of errors emitted by

```sh
# /workspace/gem5/
mypy ./src/python
```

without introducing breaking changes or causing any of the tests run by

```sh
# /workspace/gem5/tests/
./main.py run -j $(nproc) -t $(nproc)
```
to fail

I've reduced the number of errors from 706 on current stable, down to 48.

Reducing them further is certainly possible, and something I'm willing to explore (so that we can incorporate mypy into the ci system), however it may require breaking changes, or modifying things I don't understand well enough to be comfortable refactoring.

Many of the TODO's added here are more questions that I have, it may be worth addressing them before committing to merging this.

AnthonyMichaelTDM committed:
misc: fix typo in licensing info

Change-Id: I9cd6fd1ed9d91f9241a856a42c484b75f86deaaf

AnthonyMichaelTDM committed:
python: remove python2 artifacts

Change-Id: Ie27ec7d2f13e012ac968fca1fc356161e979e05f

AnthonyMichaelTDM committed:
python: explicitly handle null cases

Change-Id: Idf5f5fc3391d4472f2c7c921ce0a4e33f74a905a

AnthonyMichaelTDM committed:
python: refactor CheckedIntType __init__

Change-Id: I216a50b606a5e890578e8f3237ff1e5e7e46e191

AnthonyMichaelTDM committed:
python: refactor AddrRange __init__

this function was last updated years ago, these changes don't break
existing functionality and make the function more maintainable by using
more modern practices

Change-Id: I207b6f7f4f0ab319820dba1388d143e578fe059e

AnthonyMichaelTDM committed:
python: refactor multidict.py

use None to represent a multidict without parents instead of {} and
update the rest of the class accordingly

Change-Id: I998c307ebaf0a61849f2ce02c4c20663bed1cb8f

AnthonyMichaelTDM committed:
python: refactor MetaSimObject __init__

refactor the __init__ function to leverage the match statement in order
to more clearly handle inheritance in sim objects

behavior is unchanged, but code is more readable and therefore easier to
maintain

Change-Id: Ib4f4671f8c3ff25ae8eebbed4d50d421b74a1d0b

AnthonyMichaelTDM committed:
python: add get_mem_ranges function to AbstractBoard

Many subclasses of AbstractBoard define a mem_ranges attribute, so it's useful to have a function in the superclass that gives us access to that attribute in generic functions, while satisfying type checkers and ensuring the case where the attribute isn't defined gets properly handled

Change-Id: I0c34b1f0a1ff6b3dd9c9d8195af9f8bb174ee147

AnthonyMichaelTDM committed:
python: further improve null handling

Change-Id: I4f80bcf6a0e453459fea5fb1bd075b311d3de18f

AnthonyMichaelTDM committed:
misc: add mypy configuration to ignore missing imports

Change-Id: I6dc56eb0d8cf975dfb355d48cb286156f250752a

AnthonyMichaelTDM committed:
python: update imports, type hints, and fix typos

Change-Id: Ic26f345f8568d4dab74a81aae740e30c82625aa5

AnthonyMichaelTDM committed:
python: update imports, type hints, and fix typos. Part 2

Change-Id: I770d961e0780745cbcfd0b0ac70d540cc7d91ea9

AnthonyMichaelTDM committed:
python: patch regressions

Change-Id: Ia1d05f767efcf9ab781443a453547e7335e57c6d

AnthonyMichaelTDM commented:
it seems that while rebasing it modified one of the CI workflows? If needed I can rebase again or something

AnthonyMichaelTDM commented:
yeah, I'm gonna rebase that commit out of the tree

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -29,21 +29,25 @@
     ClockDomain,
-    IOXBar,
-    Port,
     SrcClockDomain,
-    System,
-    VoltageDomain,
+)
+from m5.objects.System import System
+from m5.objects.VoltageDomain import VoltageDomain
+from m5.objects.XBar import IOXBar
+from m5.params import (
+    AddrRange,
+    Port,
 )
```
I would rather not make these kinds of changes. It's a bit different from the traditional gem5 style.

AnthonyMichaelTDM replied:
That's fair, but without them, mypy sometimes can't tell that the items are in the objects module (or if it does, it recognizes them as modules rather than classes), the alternative would be to add `# type: ignore` to a bunch (but not all) of m5 imports

undoing these types of changes would greatly reduce the lines changes by this pr though, which itself could be reason enough to be honest.

let me know

AnthonyMichaelTDM replied:
This is mostly due to how the type stubs are generated (I think), maybe that could be something to look at more closely?

AnthonyMichaelTDM replied:
speaking of imports, are relative or absolute imports prefered?

AnthonyMichaelTDM commented:
huh, I'll look into a fix for that rq

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -101,20 +112,22 @@ def __init__(
         self.memory = memory
-        self._cache_hierarchy = cache_hierarchy
+        self._cache_hierarchy: Optional[
+            "AbstractCacheHierarchy"
+        ] = cache_hierarchy
         if cache_hierarchy is not None:
-            self.cache_hierarchy = cache_hierarchy
+            self.cache_hierarchy: "AbstractCacheHierarchy" = cache_hierarchy
```
This code seems correct, but it's really ugly. I am going to push back a bit on making changes that make the code more difficult to read.

In general, the type checking in python should improve the code's readability, not make it worse.

AnthonyMichaelTDM replied:
I'll see if there's a more readable way to write this block

AnthonyMichaelTDM replied:
yeah turns out the annotations aren't necessary in this case, will remove

AnthonyMichaelTDM replied:
fixed by [the most recent commit](https://github.com/gem5/gem5/pull/1409/commits/b73fe85e54ab679124848013b91444a7ca3aaf37)

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -133,7 +146,10 @@ def get_processor(self) -> "AbstractProcessor":
         return self.processor
 
-    def get_memory(self) -> "AbstractMemory":
+    # NOTE: Am I sure that this should be a AbstractMemorySystem instead of a m5.objects.AbstractMemory?
```
You're correct. It should be AbstractMemorySystem

AnthonyMichaelTDM replied:
good to know, thanks

AnthonyMichaelTDM replied:
I'll probably sneak removing that comment into a future commit

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -162,7 +178,13 @@ def get_cache_line_size(self) -> int:
+            return line_size
+        else:
+            # TODO: is it better to just return 0 (or some other default value) if the attribute is not set?
+            raise Exception(
```
`cache_line_size` is a parameter of `System`, so maybe this function should be in the `AbstractSystemBoard`

AnthonyMichaelTDM replied:
That makes sense, should I do that?
The reason I avoided that approach was to prevent api changes.

AnthonyMichaelTDM replied:
yeah, I did some testing (process: move the function to AbstractSystemBoard, run mypy to see what breaks, fix some of the new errors, run mypy again, ...) and ... there's a lot of things that would have to be refactored (so much so I'd probably put it in a new pr). It's definitely possible, if it makes sense it's probably a good idea, and I'm willing to do that work, but it would likely be a breaking change.

AnthonyMichaelTDM replied:
TL;DR, I think this change is worth doing, but I'd suggest we wait until after this PR is figured out and do it in a separate pr, should I open an issue?

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -318,6 +340,16 @@ def get_mem_side_coherent_io_port(self):
         raise NotImplementedError
 
+    def get_mem_ranges(self) -> List[AddrRange]:
+        if not hasattr(self, "mem_ranges"):
```
Same, this is a parameter of `System`

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -369,8 +401,8 @@ def _connect_things(self) -> None:
         # Incorporate the cache hierarchy for the motherboard.
-        if self.get_cache_hierarchy():
-            self.get_cache_hierarchy().incorporate_cache(self)
+        if cache_hierarchy := self.get_cache_hierarchy():
```
I don't like the idea of the walrus operator... call me old fashioned...

IMO, the tradeoff of slightly harder to read to save 9 characters isn't worth it.

AnthonyMichaelTDM replied:
Although I disagree, that's fair.
The alternative is to throw `self.get_cache_hierarchy()` into a temp variable and then check for None.
The point isn't to save 9 characters, it's to ensure that we're referring to the same  "cache_hierarchy" throughout the code block, the current approach (before my changes) assumes that the functions it's calling are pure functions without side-effects, and while that tends to be true it can't be statically guaranteed and could change, which would cause the code to not work as expected.

Something like

```python
cache_hierarchy = self.get_cache_hierarchy()
if cache_hierarchy:
    ...
```

powerjg commented on src/python/gem5/components/boards/abstract_system_board.py:
```diff
@@ -48,7 +56,7 @@ def __init__(
         processor: "AbstractProcessor",
         memory: "AbstractMemorySystem",
-        cache_hierarchy: "AbstractCacheHierarchy",
+        cache_hierarchy: Optional["AbstractCacheHierarchy"] = None,
```
This optional should probably be documented why it's optional somewhere...

AnthonyMichaelTDM replied:
It's optional because AbstractBoard's is optional, and some subclasses of AbstractSystemBoard want it optional (and treat it like it is)

AnthonyMichaelTDM replied:
should I add that note to a comment above the argument?

AnthonyMichaelTDM committed:
python: fix compilation error

CI fialed bc it couldn't import Self from typing, so we refacctor to not need to do that

Change-Id: I1d381bda4b2aa095a1362edcb948b49457002a00

powerjg reviewed (commented):
Wow! This is a ton of work! Thanks for the effort.

I am making some comments on the PR, though don't take these comments as action items. We should chat a bit about the right way forward.

A question for you: Do you think there's a way to do this piecemeal instead of all at once? A PR with 2000 lines changed is difficult to review.

AnthonyMichaelTDM commented:
> A question for you: Do you think there's a way to do this piecemeal instead of all at once? A PR with 2000 lines changed is difficult to review.

Yes, many of these commits aren't dependent on each other and can be arbitrarily reordered (so, can be put into their own branch and merged separately), but I'd rather avoid that for now as it makes things harder on my end.

AnthonyMichaelTDM committed:
python: remove ugly annotations in AbstractBoard.__init__

Change-Id: I2f68ebda2afcdee66b9fd05ee54c6c8677a2b8c4

AnthonyMichaelTDM commented on src/python/gem5/components/boards/kernel_disk_workload.py:
```diff
@@ -240,9 +248,13 @@ def set_kernel_disk_workload(
             if isinstance(checkpoint, Path):
-                self._checkpoint = checkpoint
-            elif isinstance(checkpoint, CheckpointResource):
-                self._checkpoint = Path(checkpoint.get_local_path())
+                self._checkpoint: Path | None = checkpoint
+            elif (
+                isinstance(checkpoint, CheckpointResource)
+                and (checkpoint_path := checkpoint.get_local_path())
+                is not None
+            ):
+                self._checkpoint = Path(checkpoint_path)
```
there are also cases like this where the walrus operator provides a lot more convenience

AnthonyMichaelTDM commented on src/python/gem5/components/processors/simple_core.py:
```diff
@@ -43,6 +42,8 @@ class SimpleCore(BaseCPUCore):
     def __init__(self, cpu_type: CPUTypes, core_id: int, isa: ISA):
         requires(isa_required=isa)
         super().__init__(
+            # WARN: we're initializing this with the wrong type, that, or the constructor expects the wrong type
+            # figure out which is the case and fix it
             core=SimpleCore.cpu_simobject_factory(
                 isa=isa, cpu_type=cpu_type, core_id=core_id
             ),
```
can you provide some insights on this?

AnthonyMichaelTDM commented on src/python/gem5/resources/resource.py:
```diff
@@ -1256,6 +1288,11 @@ def _get_to_path_and_downloader_partial(
             gem5_version=gem5_version,
             quiet=quiet,
         )
+
+    # WARN: This assertion seems to not be valid (causing some tesets to fail), investigate the cause
+    # in the meaning, we only assert when TYPE_CHECKING is True so that mypy treats this as a type hint
+    if TYPE_CHECKING:
+        assert to_path
     return to_path, downloader
```
can you provide some insight as to what was going on here? I would think that to_path would be a required argument and should not be None, but when I had this assertion it would cause some tests to fail

AnthonyMichaelTDM commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -374,9 +377,10 @@ def function_generator(func: Callable):
-        self._tick_stopwatch = []
+        self._tick_stopwatch: List[Tuple[ExitEvent, int]] = []
 
-        self._last_exit_event = None
+        # The type is actually the GlobalSimLoopExitEvent class, but that's defined in the C++ code and there are no bindings for it so I'm just using Any here.
+        self._last_exit_event: None | Any = None
```
this can be updated once/if we have better type stubs

AnthonyMichaelTDM commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -560,13 +564,22 @@ def get_last_exit_event_cause(self) -> str:
         Returns the last exit event status code
         """
-        return self._last_exit_event.getCode()
+        # TODO: determine sane default
+        return (
+            last_exit_event.getCode()
+            if (last_exit_event := self._last_exit_event)
+            else -1
+        )
```
Does -1 work as a sane default, or should something else be used?

Raising an exception is another option too because as we discussed, this shouldn't happen if people are doing things correctly

AnthonyMichaelTDM commented on src/python/m5/SimObject.py:
```diff
@@ -220,47 +232,44 @@ def __init__(cls, name, bases, dict):
-            base._instantiated = True
-        else:
-            cls._base = None
+        # TODO: "cls._values_dict" doesn't seem to exist, need to check if this is a typo
```
is _values_dict a type defined in the cpp code? if so we can blame this on the stubs and do something like
```python
if TYPE_CHECKING:
    cls._values_dict: Dict[str, Any] = getattr(cls, "_values_dict")
```
in the meantime so mypy knows what's going on

AnthonyMichaelTDM commented on src/python/m5/SimObject.py:
```diff
@@ -504,7 +513,8 @@ def py_call(self, *args, **kwargs):
             return func(self, *args, **kwargs)
 
         f = py_call if override else cxx_call
-        f.__pybind = PyBindMethod(
+        # TODO: Figure out where the __pybind attribute is coming from, and adjust type hints accordingly.
+        f.__pybind = PyBindMethod(  # type: ignore
             name,
```
just commenting here so we don't forget to address this

AnthonyMichaelTDM commented on src/python/m5/ext/pystats/simstat.py:
```diff
@@ -48,12 +48,14 @@ def __init__(
         time_conversion: Optional[TimeConversion] = None,
         simulated_begin_time: Optional[Union[int, float]] = None,
         simulated_end_time: Optional[Union[int, float]] = None,
-        **kwargs: Dict[str, Union[Group, Statistic, List[Group]]]
+        **kwargs: Dict[
+            str, Union[Group, Statistic, List[Group], List[Statistic]]
+        ],
     ):
```
note here, I don't think this is how kwargs is supposed to be typed generally, but it's iffy, this file contains a decent chunk of the remaining mypy errors

AnthonyMichaelTDM commented on src/python/m5/params.py:
```diff
@@ -587,19 +595,22 @@ def __init__(cls, name, bases, dict):
 
-        if not (hasattr(cls, "min") and hasattr(cls, "max")):
-            if not (hasattr(cls, "size") and hasattr(cls, "unsigned")):
+        if not hasattr(self, "min") or not hasattr(self, "max"):
+            if (size := getattr(self, "size", None)) is not None:
+                # TODO: Maybe only override min and max if they are not already set
+                # (rather than overriding both if either are unset)
+                if getattr(self, "unsigned", False):
+                    self.min = 0
+                    self.max = 2**size - 1
+                else:
+                    self.min = -(2 ** (size - 1))
+                    self.max = (2 ** (size - 1)) - 1
+            else:
```
Adding a comment so we can discuss this TODO, does that sound like a good idea?
It seems right, but it changes behavior, so I held off doing it

AnthonyMichaelTDM commented on src/python/m5/params.py:
```diff
@@ -897,71 +908,70 @@ def pybind_predecls(cls, code):
-                self.end = Addr(args[0])
-
-        elif len(args) == 2:
-            self.start = Addr(args[0])
-            self.end = Addr(args[1])
-        else:
-            raise TypeError("Too many arguments specified")
+                self.end = Addr(start)
+            # failure case
+            case params:
+                raise TypeError(
+                    f"Either end or size must be specified: {params}"
+                )
 
-        if kwargs:
-            raise TypeError(f"Too many keywords: {list(kwargs.keys())}")
+        # now on to the optional bit
+        if intlvMatch is not None:
+            self.intlvMatch = int(intlvMatch)
+
+        if masks is not None:
+            self.masks = [int(x) for x in list(masks)]
+            self.intlvBits = len(self.masks)
+        elif intlvBits is not None:
+            self.intlvBits = int(intlvBits)
+            self.masks = [0] * self.intlvBits
+            if intlvHighBit is None:
+                raise TypeError("No interleave bits specified")
+            intlv_high_bit = int(intlvHighBit)
+            xor_high_bit = 0
+            if xorHighBit is not None:
+                xor_high_bit = int(xorHighBit)
+            for i in range(0, self.intlvBits):
+                bit1 = intlv_high_bit - i
+                mask = 1 << bit1
+                if xor_high_bit != 0:
+                    bit2 = xor_high_bit - i
+                    mask |= 1 << bit2
+                self.masks[self.intlvBits - i - 1] = mask
 
```
I'm pretty highly confident this is a correct implementation, but I would appreciate another set of eyes

AnthonyMichaelTDM commented:
just added some comments of my own to indicate some "points of interest" for future discussion

erin-le commented:
Adding a checklist so the gem5 dev team can review this PR commit by commit:

- [x] [misc: fix typo in licensing info](https://github.com/gem5/gem5/pull/1409/commits/c704e1b0d3ede9b34d9bc9d71bef207e7bd9140e)
- [x] [python: remove python2 artifacts](https://github.com/gem5/gem5/pull/1409/commits/6ab1e15aa72e5a5cc370c030076a69e7083443c3)
- [x] [python: refactor CheckedIntType __init__](https://github.com/gem5/gem5/pull/1409/commits/a1d33274ee7e3cb9d75e046c3b872cb07c10cf9b)
    - comment from author about TODO message that still needs to be addressed
- [ ] [python: refactor AddrRange __init__](https://github.com/gem5/gem5/pull/1409/commits/12fbf13976ea666b642611730855e81fd542d249)
- [ ] [python: refactor multidict.py](https://github.com/gem5/gem5/pull/1409/commits/294e07d5abaefc06ba122a9a05041e29a86f8bd7)
- [ ] [python: refactor MetaSimObject __init__](https://github.com/gem5/gem5/pull/1409/commits/ada5a8484c22121a99dc9ec3021ce3fa57fc09c7)
- [ ] [python: explicitly handle null cases](https://github.com/gem5/gem5/pull/1409/commits/f28b706f7a833cc64988bd90618b586517d0ff80)
- [ ] [misc: add mypy configuration to ignore missing imports](https://github.com/gem5/gem5/pull/1409/commits/d3d62c5d054ad25e8a3f87d407594b7c764d105d)
- [ ] [python: update imports, type hints, and fix typos](https://github.com/gem5/gem5/pull/1409/commits/d0d91782a2180923e380a9a5893fed62c6c715d6)
- [ ] [python: add get_mem_ranges function to AbstractBoard](https://github.com/gem5/gem5/pull/1409/commits/37d2f498d5d7835e06d7df4ebe567a622cc0c462)
- [ ] [python: further improve null handling](https://github.com/gem5/gem5/pull/1409/commits/9a5d64ad655dbf8ca39ac931f5601da961e1af5f)
- [ ] [python: update imports, type hints, and fix typos. Part 2](https://github.com/gem5/gem5/pull/1409/commits/978d996cfbb62c7a994e248d45224ae0da90ff49)
- [ ] [python: patch regressions](https://github.com/gem5/gem5/pull/1409/commits/d2c6162a928de41b6825fe0c24e058448927dbe4)
- [ ] [python: fix compilation error](https://github.com/gem5/gem5/pull/1409/commits/9530079ab2a92fd3513525ccae63446a069e66e0)
- [ ] [python: remove ugly annotations in AbstractBoard.__init__](https://github.com/gem5/gem5/pull/1409/commits/b73fe85e54ab679124848013b91444a7ca3aaf37)

AnthonyMichaelTDM committed:
remove note

powerjg commented:
Thanks for this contribution. However, given the lack of activity, I am going to close this PR. If you (or anyone) wants to improve our type checking, I strongly suggest breaking the PR into smaller pieces to make reviewing easier. It may also be a good idea to add some python tests before making big refactoring changes.
